### PR TITLE
OPL-64: New calcrule - Unconditional cash payment

### PR DESCRIPTION
**Pull request by wzglinieckisoldevelo**
https://openimis.atlassian.net/browse/OPL-64

New calcrule - unconditional cash payment

### CODE REVIEW

File: calcrule_unconditional_cash_payment/calculation_rule.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                cls.signal_get_rule_details.connect(cls.get_rule_details, dispatch_uid="on_get_rule_details_signal")
+                cls.signal_get_param.connect(cls.get_parameters, dispatch_uid="on_get_param_signal")
+                cls.signal_get_linked_class.connect(cls.get_linked_class, dispatch_uid="on_get_linked_class_signal")
+                cls.signal_calculate_event.connect(cls.run_calculation_rules, dispatch_uid="on_calculate_event_signal")
+                cls.signal_convert_from_to.connect(cls.run_convert, dispatch_uid="on_convert_from_to")
+
+    @classmethod
+    def active_for_object(cls, instance, context, type="account_payable", sub_type="cash_payment"):
+        return instance.__class__.__name__ == "Policy" \
+               and context in ["policyCreated"] \
```
Comment (sniedzielski) [inline calcrule_unconditional_cash_payment/calculation_rule.py:51]: @wzglinieckisoldevelo Please replace 'policyCreated' into 'PolicyCreated'

File: calcrule_unconditional_cash_payment/calculation_rule.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    def calculate(cls, instance, **kwargs):
+        context = kwargs.get('context', None)
+        user = kwargs.get('user', None)
+        class_name = instance.__class__.__name__
+        if class_name == "Policy":
+            product = instance.product
+            payment_plan = PaymentPlan.objects.filter(benefit_plan=product, calculation=cls.uuid, is_deleted=False)
+            if payment_plan:
+                payment_plan = payment_plan.first()
+                cls.run_convert(instance=instance, context=context, convert_to='Bill', payment_plan=payment_plan,
```
Comment (sniedzielski) [inline calcrule_unconditional_cash_payment/calculation_rule.py:83]: @wzglinieckisoldevelo Please write this in such way as below:
`run_convert(
    instance=instance, 
   context=context, 
   convert_to='Bill', 
   payment_plan=payment_plan,
)
`

File: calcrule_unconditional_cash_payment/calculation_rule.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                                user=user)
+            return "conversion finished " + cls.calculation_rule_name
```
Comment (sniedzielski) [inline calcrule_unconditional_cash_payment/calculation_rule.py:85]: @wzglinieckisoldevelo Please use F string

File: calcrule_unconditional_cash_payment/calculation_rule.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    @classmethod
+    def convert(cls, instance, convert_to, **kwargs):
+        context = kwargs.get('context', None)
+        payment_plan = kwargs.get('payment_plan')
+        convert_results = {}
+        if context == "policyCreated":
+            lumpsum_to_be_paid = None
+            invoice_label = None
+            if 'calculation_rule' in payment_plan.json_ext:
```
Comment (sniedzielski) [inline calcrule_unconditional_cash_payment/calculation_rule.py:113]: @wzglinieckisoldevelo Please move 'obtaining values from json_ext' logic into separate private method. You can also as you wish move this into 'utils'.

File: calcrule_unconditional_cash_payment/calculation_rule.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                calculation_rule_json_ext_dict = payment_plan.json_ext['calculation_rule']
+                if 'lumpsum_to_be_paid' in calculation_rule_json_ext_dict:
+                    lumpsum_to_be_paid = calculation_rule_json_ext_dict['lumpsum_to_be_paid']
+                if 'invoice_label' in calculation_rule_json_ext_dict:
+                    invoice_label = calculation_rule_json_ext_dict['invoice_label']
+
+            if convert_to == 'Bill':
```
Comment (sniedzielski) [inline calcrule_unconditional_cash_payment/calculation_rule.py:120]: @wzglinieckisoldevelo You can also move this part into separate method.

File: calcrule_unconditional_cash_payment/calculation_rule.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        if class_name == "PaymentPlan":
+            list_class.append("Calculation")
+        return list_class
+
+    @classmethod
+    def convert(cls, instance, convert_to, **kwargs):
+        context = kwargs.get('context', None)
+        payment_plan = kwargs.get('payment_plan')
+        convert_results = {}
+        if context == "policyCreated":
```
Comment (sniedzielski) [inline calcrule_unconditional_cash_payment/calculation_rule.py:110]: @wzglinieckisoldevelo Please replace 'policyCreated' into 'PolicyCreated'

File: calcrule_unconditional_cash_payment/calculation_rule.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                calculation_rule_json_ext_dict = payment_plan.json_ext['calculation_rule']
+                if 'lumpsum_to_be_paid' in calculation_rule_json_ext_dict:
+                    lumpsum_to_be_paid = calculation_rule_json_ext_dict['lumpsum_to_be_paid']
+                if 'invoice_label' in calculation_rule_json_ext_dict:
+                    invoice_label = calculation_rule_json_ext_dict['invoice_label']
+
+            if convert_to == 'Bill':
+                convert_results['bill_data'] = PolicyToBillConverter.to_bill_obj(payment_plan, instance,
+                                                                                 lumpsum_to_be_paid, invoice_label)
+                convert_results['bill_data_line'] = [PolicyToBillItemConverter.to_bill_item_obj(instance,
```
Comment (sniedzielski) [inline calcrule_unconditional_cash_payment/calculation_rule.py:123]: @wzglinieckisoldevelo same as above

**Review by sniedzielski — CHANGES_REQUESTED** [review]
@wzglinieckisoldevelo Good job. Please refer to the comments. Mainly code aesthetics,

File: calcrule_unconditional_cash_payment/converters/policy_to_bill.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        bill['thirdparty_type'] = ContentType.objects.get_for_model(policy.family.head_insuree)
+
+    @classmethod
+    def build_code(cls, bill, policy):
+        bill["code"] = f"IV-UC-{policy.family.head_insuree.chf_id}-{policy.id}"
+
+    @classmethod
+    def build_price(cls, bill, lumpsum_to_be_paid):
+        bill["amount_net"] = lumpsum_to_be_paid
+
```
Comment (sniedzielski) [inline calcrule_unconditional_cash_payment/converters/policy_to_bill.py:38]: @wzglinieckisoldevelo What about 'unit_price'?

**Comment by wzglinieckisoldevelo** [comment]
I've implemented those fixes as requested,

**Review by sniedzielski — APPROVED** [review]
@wzglinieckisoldevelo thx for improving solution! Good job!

**Comment by sniedzielski** [comment]
@delcroip @wzglinieckisoldevelo ok. We can add later simple test cases (unit tests): 
- test conversion from Policy to Bill json
- test calculating amount (check if proper value is taken from json_ext PaymentPlan (lumpsum_to_be_paid))

**Pull request merged**